pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/XPathRule.java

```diff
@@ -30,6 +30,7 @@
  */
 public class XPathRule extends AbstractRule {
 
+    // TODO 7.0.0 use PropertyDescriptor<String>
     public static final StringProperty XPATH_DESCRIPTOR = StringProperty.named("xpath")
```

Comment:
Same here - the properties of XPathRule shouldn't probably be public.
Maybe tag them for now as \@InternalApi?

Reply:
In fact I think for 7.0.0 XPathRule as a whole should be internal API. Since we're introducing XML sugar for XPath rules, we could jump on the occasion to completely hide away the implementation of XPath rules, and only provide a set of interfaces as API. That would allow language implementations to use different XPath rule classes at their discretion, and us to change the implementation transparently for any users. XPathRule's API would anyway be significantly broken by dropping Jaxen...

Reply:
Yes, this makes sense, having XPathRule as whole internal API.